Spigot-Server-Patches/0406-Add-Zstd-LZ4-GZIP-write-support-for-region-files.patch

```diff
@@ -0,0 +1,1123 @@
++                        } else if (b0 == 126) {
++                            abyte = new byte[l - 1];
++                            this.b.read(abyte);
++
++                            return new DataInputStream(new BufferedInputStream(new net.jpountz.lz4.LZ4BlockInputStream(new ByteArrayInputStream(abyte), lz4Decompressor), 32 * 1024));
++                        } else if (b0 == 127) {
+                             abyte = new byte[l - 1];
+                             this.b.read(abyte);
+-                            return new DataInputStream(new BufferedInputStream(new InflaterInputStream(new ByteArrayInputStream(abyte))));
+-                        } else {
++
++                            int power = 1;
++                            byte[] outBuf;
++
++                            long decompressedBytes;
++                            boolean resize;
++                            do {
++                                resize = false;
++                                outBuf = new byte[1024 * 64 * power];
++
++                                decompressedBytes = com.github.luben.zstd.Zstd.decompressFastDict(outBuf, 0, abyte, 0, abyte.length, zstdDictDecompressor);
++                                if (com.github.luben.zstd.Zstd.isError(decompressedBytes)) {
++                                    if (com.github.luben.zstd.Zstd.getErrorCode(decompressedBytes) == com.github.luben.zstd.Zstd.errDstSizeTooSmall()) {
++                                        resize = true;
++                                    } else {
++                                        throw new com.github.luben.zstd.ZstdException(decompressedBytes);
++                                    }
++                                }
++
++                                if (resize) {
++                                    power++;
++                                }
++                            } while (resize);
++
++                            abyte = new byte[(int) decompressedBytes];
++                            System.arraycopy(outBuf, 0, abyte, 0, (int) decompressedBytes);
++
++                            return new DataInputStream(new BufferedInputStream(new ByteArrayInputStream(abyte), 32 * 1024));
++                        }
++                        // Paper end
++                        else {
+                             return null;
+                         }
+                     }
+                 }
+             }
+-        } catch (IOException ioexception) {
++        } catch (IOException | DataFormatException | com.github.luben.zstd.ZstdException ioexception) { // Paper - new & optimized compression
+             ServerInternalException.reportInternalException(ioexception); // Paper
+             return null;
+         }
+@@ -294,6 +386,7 @@ public class RegionFile implements AutoCloseable {
+ 
+             // Paper - move this into writeChunkData
+         } catch (IOException ioexception) {
++            ioexception.printStackTrace();
```

Comment:
missing //paper comment

Reply:
Totally didn't accidentally leave in debug stuff. Nope.